### PR TITLE
Rework mount namespace support

### PR DESCRIPTION
**Pull request by zyga**
This patch changes the contents of the initialized mount namespace. The
main motivator is to allow /media to be shared amongst the initial mount
namespace and the derivative namespaces created by snap-confine.

The code is unified and simplified around the new boostrap function. It
now takes a few simple arguments through a configuration structure. The
main argument is the location of the starting point of the desired root
filesystem. On classic this is the core snap (or ubuntu-core if core is
not available). On an all-snap system this is the root filesystem.

The bootstrap function puts the desired root filesystem into a new
temporary directory (using recursive bind mount) paying special
attention to avoid creating bind mount loops. This directory is now
called the constructed root filesystem and eventually, this is where we
pivot_root into.

Since we still have access to the initial root filesystem under the
initial sharing parameters as well as to the new constructed root
filesystem, we can choose to bind mount arbitrary directories over.

The code now supports a list of directories and a flag indicating if 
the mount event propagation should be bidirectional or not.

In this patch /media is on the bidirectional list. On classic a large
list of directories is on the unidirectional list. On all snap the
unidirectional list is empty because we don't need anything as we're
sharing the root filesystem which is already correctly populated by the
all-snap initrd.

The existing code for special treatment of /snap and /etc/alternatives
was retained as is (semantically) but has been folded
into the bootstrap function for simplicity. This was not yet done for /tmp
and /dev/pts.

The code for nvidia support is now also handled by the bootstrap
function but this should be moved outside later. This can be done only
after nvidia code is adjusted to assume it executes after pivot root
(this will simplify it and allow us to do something sensible in all-snap
system later). I chose not to do this to simplify the change and review
process. To repeat, nvidia support is exactly as it was before.

The apparmor profile was adjusted to take account of all the new
(numerous) operations.

Signed-off-by: Zygmunt Krynicki zygmunt.krynicki@canonical.com

### CODE REVIEW

File: src/mount-support.c
Comment (zyga) [inline src/mount-support.c]: FYI, I was split on this (advice from stgraber earlier) but it seems to actually do the right thing (to be clear, the code as-is, with `#if 0`). I didn't see any adverse effects either.

Comment (Conan-Kudo) [reply src/mount-support.c]: Please provide a comment on why `#if 0`

Comment (zyga) [reply src/mount-support.c]: I dropped that part entirely now

File: src/mount-support.c
Comment (tyhicks) [inline src/mount-support.c]: s/isolated/isolate/

File: src/mount-support.c
Comment (tyhicks) [inline src/mount-support.c]: Please also point out that it allows snaps with sufficient privs to remove mount points that were visible to the rest of the system.

File: src/mount-support.c
Comment (tyhicks) [inline src/mount-support.c]: Optional suggestion: save a little stack space and declare scratch_dir like so:

 `char scratch_dir[] = "/tmp/snap.rootfs_XXXXXX";`

File: src/mount-support.c
Comment (tyhicks) [inline src/mount-support.c]: s/over of the/over the/

File: src/mount-support.c
Comment (tyhicks) [inline src/mount-support.c]: s/applied the root/applied to the root/

File: src/mount-support.c
Comment (tyhicks) [inline src/mount-support.c:377]: I'm a little worried that bidirectional_mounts and unidirectional_mounts are two different lists and that bidirectional is also processed before unidirectional. The concern comes from the possible need to set up a unidirectional mount before a bidirectional. This could be solved by combining them into one list of something like:

```
struct sc_mounts {
    const char *path;
    bool is_bidirectional;
};
```

This is only an observation and not a blocker for this PR.

File: src/mount-support.c
Comment (tyhicks) [inline src/mount-support.c]: I think "-o slave" works but we should change "-o slave" to "--make-slave" to match the rest of the debugging output.

File: src/mount-support.c
Comment (tyhicks) [inline src/mount-support.c]: Here, too.

File: src/mount-support.c
Comment (tyhicks) [inline src/mount-support.c:407]: Add --make-rslave.

File: src/mount-support.c
Comment (tyhicks) [inline src/mount-support.c:410]: s/-o slave/--make-rslave/

File: src/mount-support.c
Comment (tyhicks) [inline src/mount-support.c]: There's an extra "slave" in there.

File: src/mount-support.c
Comment (tyhicks) [inline src/mount-support.c]: Huh, here's what I was talking about earlier in the review where I was worried about bidirectional_mounts being processed before unidirectional_mounts.

If the MERGED_USR macro is set, then the /run/media bidirectional mount is performed. The problem is that the /run unidirectional mount will then clobber it and break the sharing of /run/media across snaps and the host.

File: src/snap-confine.apparmor.in
Comment (tyhicks) [inline src/snap-confine.apparmor.in:160]: The AppArmor profile changes look incomplete. These are mostly all mounts that were already being performed and would therefore need AppArmor rules allowing them. It looks like this PR should either be reusing existing rules or removing the old rules that no longer suffice.

I'm holding off on reviewing the profile changes for now.

**Review by tyhicks — CHANGES_REQUESTED** [review]
This is a very complex change that is difficult to review in a short amount of time. However, the tests that you created are a big help in trusting the changes.

Most of my comments are suggestions and are not blockers. The only two blockers are the /run unidirectional mount happening after, and breaking, the /run/media unidirectional mount and the missing AppArmor profile changes.

File: src/mount-support.c
Comment (zyga) [reply src/mount-support.c]: Ah, correct. I was not testing this on fedora (no fedora CI yet) and that's where we use merged user. Nice catch!

File: src/mount-support.c
Comment (zyga) [reply src/mount-support.c]: This is subtly different. When you call `mount --make-slave` you get different arguments to the system call:

`mount("none", "/some/path", NULL, MS_SLAVE, NULL)`

File: src/mount-support.c
Comment (zyga) [reply src/mount-support.c:377]: I'll do this, good idea.

Comment (zyga) [reply src/mount-support.c:377]: Done, , please have a look

File: src/mount-support.c
Comment (zyga) [reply src/mount-support.c]: Fixed.

File: src/snap-confine.apparmor.in
Comment (zyga) [reply src/snap-confine.apparmor.in:160]: I've simplified the AA profile considerably. Please have a look

File: src/mount-support.c
Comment (zyga) [reply src/mount-support.c:407]: Ah, I see. Will do

Comment (zyga) [reply src/mount-support.c:407]: I take that back, --make-rslave is directly below this code

**Comment by zyga** [comment]
I ran this branch against tests in snapd and ... crashed the kernel:

```
linode:ubuntu-16.04-64 .../tests/main/snap-set# journalctl -xe
Oct 13 10:59:18 ubuntu kernel: Hardware name: QEMU Standard PC (i440FX + PIIX, 1996), BIOS rel-1.8.2-0-g33fbe13 by qemu-project.org 04/01/2014
Oct 13 10:59:18 ubuntu kernel: task: ffff880037b4d280 ti: ffff88007c344000 task.ti: ffff88007c344000
Oct 13 10:59:18 ubuntu kernel: RIP: 0010:[<ffffffff8123cc9e>]  [<ffffffff8123cc9e>] propagate_one+0xbe/0x1c0
Oct 13 10:59:18 ubuntu kernel: RSP: 0018:ffff88007c347d68  EFLAGS: 00010297
Oct 13 10:59:18 ubuntu kernel: RAX: ffff880037623a80 RBX: ffff88004f8d3480 RCX: ffff88004f8d3180
Oct 13 10:59:18 ubuntu kernel: RDX: 0000000000000000 RSI: 0000000000000076 RDI: 0000000000000000
Oct 13 10:59:18 ubuntu kernel: RBP: ffff88007c347d78 R08: ffff88007b297000 R09: ffffffff813eadcc
Oct 13 10:59:18 ubuntu kernel: R10: ffffea0001ec0c00 R11: 0000000000003f91 R12: ffff8800375f4900
Oct 13 10:59:18 ubuntu kernel: R13: ffff88007c347dc0 R14: ffff88004f8d3480 R15: 0000000000000000
Oct 13 10:59:18 ubuntu kernel: FS:  00007f595e1ca840(0000) GS:ffff88007fc00000(0000) knlGS:0000000000000000
Oct 13 10:59:18 ubuntu kernel: CS:  0010 DS: 0000 ES: 0000 CR0: 0000000080050033
Oct 13 10:59:18 ubuntu kernel: CR2: 0000000000000010 CR3: 0000000057974000 CR4: 00000000001406f0
Oct 13 10:59:18 ubuntu kernel: Stack:
Oct 13 10:59:18 ubuntu kernel:  ffff88004f8d3480 ffff8800375f4900 ffff88007c347db0 ffffffff8123d1c0
Oct 13 10:59:18 ubuntu kernel:  ffff880037623a80 ffff8800375f4900 ffff88003776e380 0000000000000000
Oct 13 10:59:18 ubuntu kernel:  ffff88007c347e98 ffff88007c347df8 ffffffff8122def7 ffff88007b0bea80
Oct 13 10:59:18 ubuntu kernel: Call Trace:
Oct 13 10:59:18 ubuntu kernel:  [<ffffffff8123d1c0>] propagate_mnt+0x120/0x150
Oct 13 10:59:18 ubuntu kernel:  [<ffffffff8122def7>] attach_recursive_mnt+0x147/0x230
Oct 13 10:59:18 ubuntu kernel:  [<ffffffff8122e038>] graft_tree+0x58/0x90
Oct 13 10:59:18 ubuntu kernel:  [<ffffffff8122e0fe>] do_add_mount+0x8e/0xd0
Oct 13 10:59:18 ubuntu kernel:  [<ffffffff8122eed0>] do_mount+0x2c0/0xe00
Oct 13 10:59:18 ubuntu kernel:  [<ffffffff8122d064>] ? mntput+0x24/0x40
Oct 13 10:59:18 ubuntu kernel:  [<ffffffff811eebf4>] ? __kmalloc_track_caller+0x1b4/0x250
Oct 13 10:59:18 ubuntu kernel:  [<ffffffff8120e5c0>] ? __fput+0x190/0x220
Oct 13 10:59:18 ubuntu kernel:  [<ffffffff811abda2>] ? memdup_user+0x42/0x70
Oct 13 10:59:18 ubuntu kernel:  [<ffffffff8122fd3f>] SyS_mount+0x9f/0x100
Oct 13 10:59:18 ubuntu kernel:  [<ffffffff818244f2>] entry_SYSCALL_64_fastpath+0x16/0x71
Oct 13 10:59:18 ubuntu kernel: Code: 39 90 d8 00 00 00 75 ec 8b b0 10 01 00 00 48 89 3d 20 e0 f8 00 48 89 05 21 e0 f8 00 39 b1 10 01 00 00 74 19 48 8b bf d8 00 00 00 <48> 8b 47 10
Oct 13 10:59:18 ubuntu kernel: RIP  [<ffffffff8123cc9e>] propagate_one+0xbe/0x1c0
Oct 13 10:59:18 ubuntu kernel:  RSP <ffff88007c347d68>
Oct 13 10:59:18 ubuntu kernel: CR2: 0000000000000010
Oct 13 10:59:18 ubuntu kernel: ---[ end trace f43e7d84ab4ddab3 ]---
Oct 13 10:59:18 ubuntu systemd[1]: snap-failing\x2dconfig\x2dhooks-x1.mount: Mount process exited, code=killed status=9
Oct 13 10:59:18 ubuntu systemd[1]: Failed to mount Mount unit for failing-config-hooks.
```

**Comment by jamiedbennett** [comment]
Looks like the problem is an outdated kernel. This was fixed in May in the Ubuntu kernels, the original fix can be found here: https://marc.info/?l=linux-fsdevel&m=146246187014403 and the LP bug https://bugs.launchpad.net/ubuntu/+source/linux/+bug/1572316

File: src/snap-confine.apparmor.in
Comment (tyhicks) [inline src/snap-confine.apparmor.in:180]: Isn't `@SNAP_MOUNT_DIR@` set to `/snap` in most situations? I don't think we need both sets of rules here.

File: src/snap-confine.apparmor.in
Comment (tyhicks) [inline src/snap-confine.apparmor.in:102]: There's another bidirectional mount when MERGED_USR is set:

  `mount options=(rw rbind) /run/media/ -> /tmp/snap.rootfs_*/run/media/,`

However, I think you said that MERGED_USR is only used on Fedora and we know that SELinux is used there instead of AppArmor so, in practice, this isn't an issue.

**Review by tyhicks — APPROVED** [review]
This gets my ack. Thanks for the fixups!

I do ask for you to clarify and fix, if needed, the profile question that I had around the potentially duplicate /snap/ and @SNAP_MOUNT_DIR@/ rules.

Thanks @zyga!

File: src/snap-confine.apparmor.in
Comment (zyga) [reply src/snap-confine.apparmor.in:180]: We need both for before pivot `@SNAP_MOUNT_DIR@/` and after pivot `/snap/`

Comment (tyhicks) [reply src/snap-confine.apparmor.in:180]: Thanks for the clarification.

File: src/snap-confine.apparmor.in
Comment (zyga) [reply src/snap-confine.apparmor.in:102]: Exactly :)

**Comment by zyga** [comment]
This now passed snapd tests in classic using https://github.com/zyga/snapd/commit/06e5942b15f86dd648f84d15db9343b55ccfee02#diff-556bb7431481e375713ea3e0883a771aR82 (just modified the branch name to test against this branch). Once I can repeat that with core (aka all-snap) I'll press the big green merge button.

**Comment by tyhicks** [comment]
Thanks for fixing the typo. Looks good to me.

**Pull request merged**
